File: 프로그래머스/3/72414. 광고 삽입/광고 삽입.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

long long adv[360010];

int time_to_seconds(string str) {
    int time = 0;
    
    time += ((str[0] - '0') * 10 + (str[1] - '0')) * 3600;
    time += ((str[3] - '0') * 10 + (str[4] - '0')) * 60;
    time += ((str[6] - '0') * 10 + (str[7] - '0'));
    
    return time;
}

string seocnds_to_time(int time){
    string h = to_string(time / 3600);
    if(h.length() == 1) h = "0" + h;
    time %= 3600;
    
    string m = to_string(time / 60);
    if(m.length() == 1) m = "0" + m;
    time %= 60;
    
    string s = to_string(time);
    if(s.length() == 1) s = "0" + s;
    
    return h + ':' + m + ':' + s;
}

string solution(string play_time, string adv_time, vector<string> logs) {
    string answer = "";

    int p_time = time_to_seconds(play_time);
    int a_time = time_to_seconds(adv_time);

    for (int i = 0; i < logs.size(); ++i) {
        int start = time_to_seconds(logs[i].substr(0, 8));
        int end = time_to_seconds(logs[i].substr(9, 8));

        adv[start]++;
        adv[end]--;
    }

    for (int i = 1; i <= p_time; ++i) {
        adv[i] += adv[i - 1];
    }
    for (int i = 1; i <= p_time; ++i) {
        adv[i] += adv[i - 1];
    }


    int start_time = -1;
    long long max_time = -1;

    for (int i = 0; i + a_time - 1 <= p_time; ++i) {
        long long sum = adv[i + a_time - 1] - (i ? adv[i - 1] : 0);
        if (sum > max_time) {
            start_time = i;
            max_time = sum;
        }
    }

    answer = seocnds_to_time(start_time);

    return answer;
}

File: 프로그래머스/3/72414. 광고 삽입/광고 삽입.java
import java.util.*;

class Solution {
    
    static long[] adv_stamp;
    
    public String solution(String play_time, String adv_time, String[] logs) {
        String answer = "";
        
        adv_stamp = new long[360001];
        
        int p_time_second = timeToSecond(play_time); // 전체 플레이 타임
        int a_time_second = timeToSecond(adv_time); // 광고 시간
        
        for(int i = 0; i < logs.length; ++i) {
            int cur_start = timeToSecond(logs[i].substring(0, 8));
            int cur_end = timeToSecond(logs[i].substring(9, 17));

            adv_stamp[cur_start]++;
            adv_stamp[cur_end]--;
        }

        // 각 시간대별 시청자 수
        for(int i = 1; i <= p_time_second; ++i) {
            adv_stamp[i] += adv_stamp[i-1];
        }
        
        // 누적 시청자 수
        for(int i = 1; i <= p_time_second; ++i) {
            adv_stamp[i] += adv_stamp[i-1];
        }

        long  max_value = adv_stamp[a_time_second - 1];
        int start_time = 0;
        
        for(int i = 1; i + a_time_second - 1 <= p_time_second; ++i) {
            long v = adv_stamp[i + a_time_second - 1] - adv_stamp[i - 1];
            
            if(v > max_value) {
                max_value = v;
                start_time = i;
            }
        }
        
        answer = secondToTime(start_time);
        
        return answer;
    }
    
    static int timeToSecond (String time) {
        String hour = time.substring(0, 2);
        String min = time.substring(3, 5);
        String sec = time.substring(6, 8);
        
        int ret = Integer.parseInt(hour) * 3600 + Integer.parseInt(min) * 60 + Integer.parseInt(sec);

        return ret;
    }
    
    static String secondToTime (int time) {
        int hour = time / 3600;
        int min = (time % 3600) / 60;
        int sec = time % 60;
        
        String ret = String.format("%02d:%02d:%02d", hour, min, sec);
        
        return ret;
    }
}